matron/src/device/device_midi.c

```diff
@@ -86,93 +86,301 @@ int dev_midi_init(void *self, unsigned int port_index, bool multiport_device) {
     return 0;
 }
 
+int dev_midi_virtual_init(void *self) {
+    
+    struct dev_midi *midi = (struct dev_midi *)self;
+    struct dev_common *base = (struct dev_common *)self;
+
+    if (snd_rawmidi_open(&midi->handle_in, &midi->handle_out, "virtual", 0) < 0) {
+        fprintf(stderr, "failed to open alsa virtual device.\n");
+        return -1;
+    }
+
+    // trigger reading
+    snd_rawmidi_read(midi->handle_in, NULL, 0);
+
+    base->start = &dev_midi_start;
+    base->deinit = &dev_midi_deinit;
+
+    return 0;
+}
+
 void dev_midi_deinit(void *self) {
     struct dev_midi *midi = (struct dev_midi *)self;
     snd_rawmidi_close(midi->handle_in);
     snd_rawmidi_close(midi->handle_out);
 }
 
+/*
+ *  MIDI PARSER based from https://github.com/FluidSynth/fluidsynth/
+ * 
+ *  midi_event_type was based from fluid_midi_event_type
+ *  midi_event_t was based from fluid_midi_event_t
+ *  midi_parser_t was based from fluid_midi_parser_t 
+ *  All the above originals can be found at: https://github.com/FluidSynth/fluidsynth/blob/master/src/midi/fluid_midi.h
+ * 
+ * 
+ *  midi_parser_parse was based on fluid_midi_parser_parse 
+ *  midi_event_length was based on fluid_midi_event_length
+ *  All the above originals can be found at: https://github.com/FluidSynth/fluidsynth/blob/master/src/midi/fluid_midi.c
+ *   
+ */
```

Comment:
as discussed in the other PR, please move this code to a separate file with proper licensing and attribution.

Reply:
`matron/src/device/device_midi_parser.h` and `matron/src/device/device_midi_parser.c` would be ok?

Reply:
yeah, please also include the original copyright and license as this is a requirement of LGPL.

Reply:
Should I copy the entire [Licence](https://github.com/FluidSynth/fluidsynth/blob/master/LICENSE) on the header file (Ex: `matron/src/device/device_midi_parser.h` )?

Reply:
just the license statement will suffice: https://github.com/FluidSynth/fluidsynth/blob/master/src/midi/fluid_midi.c#L3-L19